### ISSUE TITLE
How suspend image mixing

### ISSUE THREAD

**Issue by Borda** [ultralytics/yolov5#931]
## ❔Question

I want to run image training with quite a conservative augmentation so just flipping and small translation/zoom, but the samples are always quite wild... I have set the following parameters in hyp config:
```yaml
degrees: 0.0  # image rotation (+/- deg)
translate: 0.0  # image translation (+/- fraction)
scale: 0.0  # image scale (+/- gain)
shear: 0.0  # image shear (+/- deg)
perspective: 0.0  # image perspective (+/- fraction), range 0-0.001
mixup: 0.0  # image mixup (probability)
```
but still getting these batches...
![train_batch1](https://user-images.githubusercontent.com/6035284/92476754-bed73f80-f1df-11ea-8ed0-23fa7d48306c.jpg)

**Comment by Borda** [comment]
ok, we found the option `--rect` referred in #55 but I would say that would be more logical to have the normal images by default and explicitly allow mosaic training with e.g. `--mosaic`
![train_batch2](https://user-images.githubusercontent.com/6035284/92478595-a7e61c80-f1e2-11ea-9c73-6930276d80ad.jpg)
I would also assume that some augmentations like `translate` do not have an effect in mosaic mode, right?

**Comment by glenn-jocher** [comment]
@Borda hi! All of the augmentations are applied as per your yaml file, regardless of whether mosaic is on or not. So rotation, scale, etc affect mosaic or --rect images identically.

We could add a mosaic probability in the hyp file, i.e. in the range of 0-1 like the mixup probability to make the option more uniform. Mosaic is on by default as it seems to help most datasets.

--rect is more unique. Besides the obvious difference, it sorts the entire datset by aspect ratio and then presents batches of similar aspect ratio images togethor.

**Comment by Borda** [comment]
> @Borda hi! All of the augmentations are applied as per your yaml file, regardless of whether mosaic is on or not. So rotation, scale, etc affect mosaic or --rect images identically.

ok, but how would you differ with /without translation if you don't see any complete image anyway?

> We could add a mosaic probability in the hyp file, i.e. in the range of 0-1 like the mixup probability to make the option more uniform. Mosaic is on by default as it seems to help most datasets.

yes, that would be much easier to find... also in such case the switch can be part of the hyper-param search

> --rect is more unique. Besides the obvious difference, it sorts the entire datset by aspect ratio and then presents batches of similar aspect ratio images together.

not sure it is really needed to group images in batch according to a similar scale, intuitively I would say it could be rather contra productive...

**Comment by glenn-jocher** [comment]
Mosaic includes it's own translation to force crops, which is an intentioanl design point. If an image is retranslated back to the origin, it may have lost some of it's edge content this way (a desired outcome).

I'll add mosaic to the hyps. TODO: Add hyp['mosaic'] probability parameter that acts like hyp['mixup'].

Sorting by aspect ratio allows for FLOPS reduction and speed increases when passing through the entire dataset. You can see the effect in the first batch when running `python test.py --data coco.yaml`. A negative consequence for training is lack of shuffling.
![test_batch0_gt](https://user-images.githubusercontent.com/26833433/92537208-6f4d3f80-f1f0-11ea-98cb-7d19efbc8211.jpg)

**Comment by glenn-jocher** [comment]
TODO: Add hyp['mosaic'] probability parameter that acts like hyp['mixup'].

**Comment by glenn-jocher** [comment]
@Borda I've added mosaic to the hyperparameters now in https://github.com/ultralytics/yolov5/commit/f1c63e2784dffc9c2ccacb9c5a2e753bacf011ab, and I've also taken the opportunity to add warmup parameters to the hyps at the same time.

You can vary mosaic probability now from 0-1, same as mixup, lr-flips etc.

**Comment by Borda** [comment]
just looking at the commit, do you have some description for `warmup_*` because it is not much intuitive...

**Comment by glenn-jocher** [comment]
@Borda warmup slowly ramps training parameters from their initial (more stable) values to their default training values. It's common to ramp learning rate for example from 0 to some initial value over the first few epochs to avoid early training instabilities, nan's, etc.

**Comment by glenn-jocher** [comment]
Warmup effects are visible in Tensorboard learning rate plots, which are automatically tracked since a recent commit. Example below shows about 30 epochs of warmup on a custom dataset, one plot per parameter group. Last plot shows different warmup strategies (i.e. different hyps).

<img width="847" alt="Screen Shot 2020-09-16 at 12 58 09 PM" src="https://user-images.githubusercontent.com/26833433/93386308-578f4000-f81c-11ea-8a40-4dbfa8a0a51c.png">

**Comment by Borda** [comment]
@glenn-jocher sure, I was more-less meaning add some better comments to the configs or/and add an explanation in wiki/readme
